### ISSUE TITLE
Remove asyncore, asynchat and smtpd modules

### ISSUE THREAD

**Issue by Mariatta** [python/cpython#72719]
BPO | [28533](https://bugs.python.org/issue28533)
--- | :---
Nosy | @warsaw, @orsenthil, @vstinner, @asvetlov, @1st1, @grzgrzgrz3, @Mariatta, @nanjekyejoannah, @aeros, @jacksonriley, @OhSoDank, @iritkatriel
PRs | <li>python/cpython#29521</li><li>python/cpython#29951</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2016-10-25.18:32:04.420>
labels = ['3.11', 'tests', 'expert-asyncio']
title = 'Remove asyncore, asynchat and smtpd modules'
updated_at = <Date 2022-02-24.13:39:30.092>
user = 'https://github.com/Mariatta'
```

bugs.python.org fields:
```python
activity = <Date 2022-02-24.13:39:30.092>
actor = 'vstinner'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Tests', 'asyncio']
creation = <Date 2016-10-25.18:32:04.420>
creator = 'Mariatta'
dependencies = []
files = []
hgrepos = []
issue_num = 28533
keywords = ['patch']
message_count = 37.0
messages = ['279439', '291907', '291908', '294710', '294712', '294771', '294773', '356355', '356364', '356369', '357088', '379342', '396295', '396297', '396319', '396323', '404575', '406159', '406160', '406161', '406162', '406170', '406173', '406208', '406210', '406213', '406214', '406215', '406372', '406377', '406434', '407559', '407560', '407897', '407900', '407915', '413909']
nosy_count = 12.0
nosy_names = ['barry', 'orsenthil', 'vstinner', 'asvetlov', 'yselivanov', 'grzgrzgrz3', 'Mariatta', 'nanjekyejoannah', 'aeros', 'jacksonriley', 'dankreso', 'iritkatriel']
pr_nums = ['29521', '29951']
priority = 'normal'
resolution = None
stage = 'resolved'
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue28533'
versions = ['Python 3.11']
```

</p></details>

* PR: gh-96580

**Comment by vstinner** [comment]
> asyncore is still used in several tests and should be replaced.

Here is the list of Python files using asyncore:

haypo@selma$ grep -l asyncore $(find -name "\*.py")
./Lib/asyncore.py
./Lib/asynchat.py
./Lib/test/test_smtpd.py
./Lib/test/test_asyncore.py
./Lib/test/test_smtplib.py
./Lib/test/test_os.py
./Lib/test/test_ssl.py
./Lib/test/test_logging.py
./Lib/test/libregrtest/save_env.py
./Lib/test/test_asynchat.py
./Lib/test/test_pyclbr.py
./Lib/test/test_support.py
./Lib/test/test_poplib.py
./Lib/test/test_ftplib.py
./Lib/smtpd.py

smtpd has been rewritten with asyncio:
http://aiosmtpd.readthedocs.io/

The documentation of the smtpd module already points to aiosmtpd.

libregrtest/save_env.py only checks that asyncore.socket_map is not modified, we can keep it until asyncore is really removed.

test_support.py only tests that it's possible to have a "clean" import of asyncore, it can be easily be replaced later.

The list can reduced to:

./Lib/test/test_smtplib.py
./Lib/test/test_os.py
./Lib/test/test_ssl.py
./Lib/test/test_logging.py
./Lib/test/test_pyclbr.py
./Lib/test/test_poplib.py
./Lib/test/test_ftplib.py

**Comment by vstinner** [comment]
asyncore was modified to emit a DeprecationWarning:
http://bugs.python.org/issue25002#msg279417

But then a lot of code starts with to fail -Werror.

Copy of Martin Panter's [msg279469](https://bugs.python.org/msg279469):

I normally run the tests with -Werror, and the failures I get are:

* test_ssl, test_smtplib, test_poplib, test_logging, test_ftplib, test_support: tests use asyncore
* test_os: test uses asynchat (When you import asynchat, the first error complains about importing asyncore, there are actually two warnings)
* test_all: This seems to ignore DeprecationWarning; perhaps an exception for PendingDeprecationWarning should also be added?
* test_asyncore and test_asynchat: Obviously these have to still test the modules, so they should anticipate the warnings, perhaps using Serhiy’s code
* test_smtpd: smtpd module itself uses asyncore; see bpo-25008

**Comment by 6dc2ace7-63c1-4dd4-a1e8-050515409c65** [comment]
I would like to work on this issue.

I think it's a good idea to split this task into few parts/PR.

Let me start from ./Lib/test/test_poplib.py.

What about rewriting pop3 server stub using asyncio, i think requests could be handled synchronously, there will be no benefits from asynchronous.

Please let me know what you think.

**Comment by vstinner** [comment]
> I think it's a good idea to split this task into few parts/PR. Let me start from ./Lib/test/test_poplib.py.

Once you have a patch, open a new issue and mention it here.

**Comment by 6dc2ace7-63c1-4dd4-a1e8-050515409c65** [comment]
./Lib/test/test_poplib.py
sub-issue bpo-28533

**Comment by 6dc2ace7-63c1-4dd4-a1e8-050515409c65** [comment]
./Lib/test/test_poplib.py
sub-issue bpo-30514

Fixed issue number from previous comment

**Comment by hugovk** [comment]
Follow-up: PEP 594 has been accepted and asyncore, asynchat and smtpd are set for removal in 3.12.

Merged documentation PRs:

* https://github.com/python/cpython/pull/31891
* https://github.com/python/cpython/pull/31984

Pending removal PR:

* https://github.com/python/cpython/pull/23728

**Comment by vstinner** [comment]
> Reported upstream (it's a single file, used for tests): https://github.com/ansible/ansible/issues/76452

Oh, I reported the issue to the wrong GitHub project. I reported it again at: https://github.com/ansible-collections/community.general/issues/4656

**Comment by vstinner** [comment]
> mercurial-6.0

I notified the project upstream: https://bz.mercurial-scm.org/show_bug.cgi?id=6700

**Comment by Karlson2k** [comment]
The [suggested stmpd replacement](https://github.com/aio-libs/aiosmtpd) supports [Python versions up to 3.9](https://github.com/aio-libs/aiosmtpd#supported-testenvs). Python 3.10 and later is not supported.
There is [no single commit in the past 14 months in the repo](https://github.com/aio-libs/aiosmtpd/commits/master).

**Comment by iritkatriel** [comment]
> Follow-up: PEP 594 has been accepted and asyncore, asynchat and smtpd are set for removal in 3.12.
> 
> Merged documentation PRs:
> 
> * [bpo-47022: Document asynchat, asyncore and smtpd removals in 3.12 #31891](https://github.com/python/cpython/pull/31891)
> * [bpo-47061: document module deprecations due to PEP 594 #31984](https://github.com/python/cpython/pull/31984)
> 
> Pending removal PR:
> 
> * [bpo-42611: PEP 594 - Remove asyncore, asynchat, and smptd from stdlib #23728](https://github.com/python/cpython/pull/23728)
> 
> Newer removal PR that was merged...
> 
> * [bpo-28533: Remove asyncore, asynchat, smtpd modules #29521](https://github.com/python/cpython/pull/29521)
> 
> And reverted:
> 
> * [Revert "bpo-28533: Remove asyncore, asynchat, smtpd modules (GH-29521)" #29951](https://github.com/python/cpython/pull/29951)



We should be able to un-revert this now.

**Comment by gvanrossum** [comment]
I propose to remove the `expert-asyncio` label from this issue. Any objections? This doesn't require an asyncio expert at all, it's just documentation and removals (or deprecations?).

**Comment by iritkatriel** [comment]
I think that was referring to tests that need to be rewritten with asyncio. The alternative is to move these modules (or parts of them) to test.support so the tests continue to work.

**Comment by gvanrossum** [comment]
> I think that was referring to tests that need to be rewritten with asyncio. The alternative is to move these modules (or parts of them) to test.support so the tests continue to work.

That still doesn't require pinging the asyncio experts, IMO (and my opinion counts since I'm the only active one :-). I'll remove the label.

**Comment by vstinner** [comment]
> The alternative is to move these modules (or parts of them) to test.support so the tests continue to work.

IMO that's a convenient way to unblock the issue. Moving away from asyncore/asynchat would be nice but nobody managed to update tests in 10 years, so maybe it's ok to not changed them until it becomes a major (blocking) issue :-)

**Comment by sobolevn** [comment]
I've moved these modules to `test/support` as @iritkatriel and @vstinner proposed.
For now, they were coppied as-is (probably some parts are unused, but this is a subject to another piece of work).

Tests do continue to pass! 🎉

**Comment by vstinner** [comment]
asyncore, asynchat, smtpd are removed again in the main branch (future Python 3.12). Let's see how it goes this time.

Summary of this issue.

The asyncore/asynchat deprecation has a long history starting in 2015: https://bugs.python.org/issue25002

In 2020, Kyle Stanley @aeros worked on the removal in a branch: https://github.com/aeros/cpython/tree/remove-asycore-asynchat-smtpd

In 2021, I already removed the 3 modules (asyncore, asynchat, smtpd) in Python 3.11, but the Steering Council asked me to revert it since these modules only emitted a DeprecationWarning in 1 Python release (Python 3.10):

* PR: https://github.com/python/cpython/pull/29521 (approved by 5 core devs)
* commit: https://github.com/python/cpython/commit/9bf2cbc4c498812e14f20d86acb61c53928a5a57
* I [announced the removal on python-dev](https://mail.python.org/archives/list/python-dev@python.org/thread/LZOOLX5EKOITW55TW7JQYKLXJUPCAJB4/)
* In December 2021, [21 projects of PyPI top 5000 projects](https://github.com/python/cpython/issues/72719#issuecomment-1093730152) used these 3 removed modules
* Brett Cannon wrote on python-dev that "the modules got yanked prematurely", the removal didn't follow PEP 387 process (updated to require a deprecation in at least 2 Python versions, rather than 1).
* I formally requested a SC exception: [python/steering-council#86](https://github.com/python/steering-council/issues/86)
* [SC rejected my exception](https://github.com/python/steering-council/issues/86#issuecomment-987412171)
* I created PR #29951 to revert my removal

After my revert, I notified Mercurial and Ansible that these 3 projects are deprecated since 2015.

In 2022, these modules are removed again in Python 3.12:

* smtpd: PR https://github.com/python/cpython/pull/93246 and commit https://github.com/python/cpython/commit/56d16e8cb4e7d7ab47efdec08a0dae8d21b155f5
* asyncore, asynchat: PR #96580 and commit https://github.com/python/cpython/commit/e56e33d271e511e7c2324640d7f49d39b49830d8

The removal is now covered by [PEP 594](https://peps.python.org/pep-0594/) which has been approved the SC.

You can now help the remaining projects still using the removed projects to be upgrade to asyncio and aiosmptd. People can also collaborate to maintain the 3 removed modules outside Python (on PyPI) if they want. Just copy code, tests and doc from Python 3.11.
